### ISSUE TITLE
Model is not being trained

### ISSUE THREAD

**Issue by jayendra13** [dnouri/nolearn#60]
I am trying to train the 'Convolutional Neural Network' by following this [tutorial](http://danielnouri.org/notes/2014/12/17/using-convolutional-neural-nets-to-detect-facial-keypoints-tutorial/)
but while running the code it seems that no actual training is being done. You can see my code [here](https://gist.github.com/jayendra13/fae78568ebff5f92aec6).
Here is a log of code for first few epochs.

```
%run train.py
... loading data
float32
(35126L, 1L, 100L, 100L)
(35126L,)
  InputLayer            (None, 1, 100, 100)     produces   10000 outputs
  Conv2DLayer           (None, 1, 98, 98)       produces    9604 outputs
  MaxPool2DLayer        (None, 1, 49, 49)       produces    2401 outputs
  Conv2DLayer           (None, 1, 48, 48)       produces    2304 outputs
  MaxPool2DLayer        (None, 1, 24, 24)       produces     576 outputs
  DenseLayer            (None, 50)              produces      50 outputs
  DenseLayer            (None, 5)               produces       5 outputs

 Epoch  |  Train loss  |  Valid loss  |  Train / Val  |  Valid acc  |  Dur
--------|--------------|--------------|---------------|-------------|-------
     1  |         nan  |         nan  |          nan  |     73.50%  |  93.6s
     2  |         nan  |         nan  |          nan  |     73.50%  |  93.2s
     3  |         nan  |         nan  |          nan  |     73.50%  |  109.9s
     4  |         nan  |         nan  |          nan  |     73.50%  |  93.6s
     5  |         nan  |         nan  |          nan  |     73.50%  |  88.5s
```
I can't figure out what I am doing wrong

**Comment by dnouri** [comment]
Is this a classification problem?  How does the network definition look like?

**Comment by jayendra13** [comment]
Yes it is a classification problem, my input is 100x100 gray-scale images which has to be classified in to 5 classes.  I have shared my [code](https://gist.github.com/jayendra13/fae78568ebff5f92aec6).
Here is a architecture of network.

```
	net = NeuralNet(
		layers=[
			('input', layers.InputLayer),
			('conv1', Conv2DLayer),
			('pool1', MaxPool2DLayer),
			('conv2', Conv2DLayer),
			('pool2', MaxPool2DLayer),
			('hidden4', layers.DenseLayer),
			('output', layers.DenseLayer),
			],
		input_shape=(None, 1, 100, 100),
		conv1_num_filters=1, conv1_filter_size=(3, 3), pool1_ds=(2, 2),
		conv2_num_filters=1, conv2_filter_size=(2, 2), pool2_ds=(2, 2),		
		hidden4_num_units=50,
		output_num_units=5, output_nonlinearity=None,
 
		update_learning_rate=0.01,
		update_momentum=0.9,
 
		regression=False,
		max_epochs=5,
		verbose=1)
```

The dimensions and dtypes of X and Y which I pass to fit method are
```
net.fit(X, Y)
```
shape of X (35126L, 1L, 100L, 100L)
shape of Y (35126L,)
dtype of  X = float32
dtype of  Y = int32

and
theano.config.floatX = float32

**Comment by paipai880429** [comment]
@jayendra13 You just mentioned that this is a classification problem, right ? Then,
```output_nonlinearity=None``` should be replaced by ```output_nonlinearity=lasagne.nonlinearities.softmax```

**Comment by jayendra13** [comment]
@paipai880429 Thanks for your comment I changed output_nonlinearity to softmax from the None, still
I am getting the same results, I think still there is something where I am making silly mistake.
One thing I forgot to mention that whenever I use predict method, it returns all zeros

**Comment by paipai880429** [comment]
@jayendra13 It seems that the learning process has never been started. Then, did you try to decrease your initial learning rate?

**Comment by jayendra13** [comment]
That might not be issue because initial random weights should give me some non-nan values for train and validation losses

**Comment by pattysan** [comment]
Hello I have the same problem. How did you fix it? Thank you

**Comment by pattysan** [comment]
I solved my problem, there were a set of value equal to inf ( it was an error) in my dataset.

**Comment by Tinmard** [comment]
Had the same issue and "softmax" fixed it for me ! 
Thks :+1:

**Comment by ramalhoruben** [comment]
@paipai880429 In my case I had a learning rate of 0.01, changing it to 0.001 solved the problem, thanks for the tip. Why would the learning rate lead to this numerical instability?! Is it related to dataset size? Mine is big. Thanks
